### DISCUSSION TITLE
Full GPT2 CORE Calibration on single RTX 5090 in 26hrs

### DISCUSSION THREAD

**Discussion by TeddyAndres** [karpathy/nanochat#754] [General]
I've been experimenting with a custom sparse vocabulary method and was able to reach full GPT-2 CORE calibration on a single RTX 5090.

Runs:

d24 - 32 hrs - CORE 0.267 - peak VRAM 25.4GB
d14 - 33.6 hrs - CORE 0.2568 - peak VRAM 28.3GB

Logs are attached.

Let me know if you have any questions.

[d14 under 34hrs GPT2 CORE.log](https://github.com/user-attachments/files/27979615/d14.under.34hrs.GPT2.CORE.log)
[d24 32hrs GPT2 CORE.log](https://github.com/user-attachments/files/27979617/d24.32hrs.GPT2.CORE.log)

**Comment by svlandeg** [comment]
Oh wow, that's cool! Which commands did you run exactly (settings for training etc) and what other edits did you make to the code? Do you have a branch/commit on your fork you can point to?

**Comment by TeddyAndres** [comment]
The run commands were:

d24: -m scripts.base_train --sparse-mode --depth 24 --window-pattern L --fp8 --total-batch-size 262144 --device-batch-size 4 --num-iterations 20000 --sparse-manifest manifests/65kvocab_2kseq_4batch_32accum_20kstep.json --warmup-ratio 0 --warmdown-ratio 0.65 --final-lr-frac 0.1 --log-every 10 --core-metric-every 5000 --run "d24 dynamic false 20ksteps 262keffbatch"

d14: -m scripts.base_train --sparse-mode --window-pattern L --fp8 --token-cache-dir  --token-cache-workers 8 --depth=14 --aspect-ratio 128 --total-batch-size 524288 --max-seq-len 2048 --device-batch-size=8 --num-iterations=12500 --run "d14 65kvoc 2kseq 8batch 32accum 128aspect 12.5kstep unembedlr0.005 0warmup 0.65warmdown 0.2embedlr 0.01matrix 0.27scalar 0wd 0.7adam1 0.95adam2" --sparse-manifest manifests/65kvocab_2kseq_8batch_32accum_15kstep.json --warmup-ratio 0 --warmdown-ratio 0.65 --final-lr-frac 0.10 --embedding-lr 0.2 --unembedding-lr=0.005 --matrix-lr 0.01 --scalar-lr 0.27 --weight-decay 0 --adam-beta1 0.7 --adam-beta2 0.95 --log-every 10 --eval-every 250 --core-metric-every 2500 --save-every 5000

I made a number of changes to the code because I have tried a lot of different experiments but the core idea is simple. The core idea is:
- evaluate the training data prior to loading the data into the model and identify the unique tokens in the training step; load only rows for those unique tokens into VRAM for the vocab-dimensioned tables (e.g. embeddings, lm_head); master tables for vocab dimensioned tables stay in RAM and they never get loaded into VRAM; swap rows each training step
- I decided on using a manifest to identify unique tokens in advance of the start of the run to benefit runtime
- in addition I tried a number of experiments but in the end the simple idea was just as good as the complicated ones
- experiments:
   - cold bias: count the number of training steps a token was out of the model during training and apply a logit adjustment scaled on the number of cold steps; this was effective in reducing val BPB and ECE; tries to replicate some regularisation
   - cold/warm clouds: in addition to strictly present uniques each step add a selection of 'likely but not right' tokens (warm cloud) and 'globally likely but not the right context' (cold cloud); I didn't find many meaningful benefits with this one but I still believe the concept has potential because it's turning the idea of arbitrary regularisation from the vocabulary into a tunable dial
   - sparse top-k: evaluate the exact tokens and token pairs with the highest loss during training and insert the incorrect token into training every time the correct but not chosen token appears in the training data; I didn't find any obvious benefits with this one either
   
Some notable observations:
- at the standard nanochat d6 size the sparse vocab model uses 5GB of VRAM vs 15GB for the standard model
- at 262k effective training tokens per step the maximum number of unique tokens per step is around 25k (vs 65k for the full vocab); this saving is the source of the benefits
- for the full GPT2 CORE run it took the sparse vocab model 5.2B training tokens (a 6.72 scaling params ratio); I am interpreting this as due to higher signal to noise ratio in the softmax calcs
- this doesn't directly benefit the nanochat speed run but this sparse vocab design allows for vocabulary size increases and the model compute / footprint scales with unique tokens rather than vocab size; this means large vocabs are materially less costly to run

Let me know if there are any more questions.

**Comment by svlandeg** [comment]
Cool! Care to make your fork public so others can try it out, too?

**Comment by TeddyAndres** [comment]
It's public now. This is probably the best place to start: https://github.com/TeddyAndres/nanochat/tree/feature/upstream-sparse-runtime

Things to note:
- the model in its current form relies on a manifest to identify the unique tokens per step
- there is a script to create the manifest in scripts
- the dimensions of the manifest need to match the batch/accum/ddp dimensions of the run and have at least as many training steps
- the manifest is set to use pre-tokenized data so the script will create a cache if one does not already exist

There are some more details here:
docs/sparse-manifest-run.md

Good luck and happy training!

**Comment by TeddyAndres** [comment]
There's now an updated best time to GPT2 CORE on a single RTX 5090. I got CORE = 0.2569 under 26hrs. Details:

- D22
- seq len = 2048
- batch = 6
- grad accum = 26
- steps = 17k
- final val BPB = 0.7317
- CORE = 0.2569
[d22 under 26 hours GPT2 CORE.log](https://github.com/user-attachments/files/28164950/d22.under.26.hours.GPT2.CORE.log)
